From 6e4a72e8d83a904454578d29365134e461596fba Mon Sep 17 00:00:00 2001
From: Markos Chandras <markos.chandras@imgtec.com>
Date: Tue, 03 Dec 2013 14:01:31 +0000
Subject: MIPS: futex: Use LLE and SCE for EVA

For EVA, We should use the LL/SC equivalents which are
LLE and SCE to safely perform a futex operation.

Signed-off-by: Markos Chandras <markos.chandras@imgtec.com>
---
(limited to 'arch/mips/include/asm/futex.h')

diff --git a/arch/mips/include/asm/futex.h b/arch/mips/include/asm/futex.h
index 1ce3671..f8c3a09 100644
--- a/arch/mips/include/asm/futex.h
+++ b/arch/mips/include/asm/futex.h
@@ -109,13 +109,11 @@
 		"	.set	push				\n"	\
 		"	.set	noat				\n"	\
 		"       .set    eva                             \n"     \
-		"1:     lwe     %1, %4                          \n"     \
-		"       lwe     %1, %4                          \n"     \
+		"1:     lle     %1, %4                          \n"     \
 		"	.set	mips0				\n"	\
 		"	" insn	"				\n"	\
 		"       .set    eva                             \n"     \
-		"2:     swe     $1, %2                          \n"     \
-		"       li      $1, 1                           \n"     \
+		"2:     sce     $1, %2                          \n"     \
 		"	beqz	$1, 1b				\n"	\
 		__WEAK_LLSC_MB						\
 		"3:						\n"	\
@@ -311,14 +309,12 @@ futex_atomic_cmpxchg_inatomic(u32 *uval, u32 __user *uaddr,
 		"	.set	push					\n"
 		"	.set	noat					\n"
 		"       .set    eva                                     \n"
-		"1:     lwe     %1, %3                                  \n"
-		"       lwe     %1, %3                                  \n"
+		"1:     lle     %1, %3                                  \n"
 		"       bne     %1, %z4, 3f                             \n"
 		"	.set	mips0					\n"
 		"       move    $1, %z5                                 \n"
 		"       .set    eva                                     \n"
-		"2:     swe     $1, %2                                  \n"
-		"       li      $1, 1                                   \n"
+		"2:     sce     $1, %2                                  \n"
 		"	beqz	$1, 1b					\n"
 		__WEAK_LLSC_MB
 		"3:							\n"
--
cgit v0.9.2
